Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb

```diff
@@ -2,119 +2,300 @@
+      rescue StandardError => e
+        if(e.type == 'LeaseAlreadyPresent')
+            if (retried > retry_times)
+                raise
+            end
+            retried += 1
+            sleep interval_sec
         end
       end
-    end
-    return NIL
-  end # def lock_blob
-  
-  def archive_blob(blob_names)
-    real_blob_names = blob_names.select { |name| !name.end_with?(".bak") }
-    real_blob_names.each do |blob_name|
-      new_blob_name = "#{blob_name}-#{Time.now.to_f}.bak"
-      @azure_blob.acquire_lease(@container, blob_name,{:duration=>60, :timeout=>10, :proposed_lease_id=>SecureRandom.uuid})
-      @azure_blob.copy_blob(@container, new_blob_name, @container, blob_name)  
-      @azure_blob.delete_blob(@container, blob_name)
-      return new_blob_name
-    end
-    return NIL
-  end # def lock_blob
- 
-  def process(output_queue)
-    blob_names = list_blob_names
-    
-    if @backupmode
-      blob_name = archive_blob(blob_names)
-    else
-      blob_name = lock_blob(blob_names)
-    end
-
-    return if !blob_name
-    blob, content = @azure_blob.get_blob(@container, blob_name)
-    
-
-    @codec.decode(content) do |event|
-      decorate(event)
-      output_queue << event
-    end
-  rescue LogStash::ShutdownSignal => e
-    raise e
-  rescue => e
-    @logger.error("Oh My, An error occurred.", :exception => e)
-  end # def process
-  
-  public
-  def run(output_queue)
-    while !stop?
-      process(output_queue)
-    end # loop
-  end # def run
- 
-  public
-  def teardown
-  end # def teardown
-end # class LogStash::Inputs::Azuretopic
+    end #while
+    return lease
+  end # acquire_lease
+
+  # Return the next blob for reading as well as the start index.
+  def register_for_read
+    begin
+      all_blobs = list_all_blobs
+      registry = all_blobs.find { |item| item.name.downcase == @registry_path  }
+      registry_locker = all_blobs.find { |item| item.name.downcase == @registry_locker }
+
+      candidate_blobs = all_blobs.select { |item| (item.name.downcase != @registry_path) && ( item.name.downcase != @registry_locker ) }
+      
+      start_index = 0
+      lease = nil;
+
+      # Put lease on locker file than the registy file to allow update of the registry as a workaround for Azure Storage Ruby SDK issue # 16.
+      # Workaround: https://github.com/Azure/azure-storage-ruby/issues/16
+      registry_locker = @azure_blob.create_block_blob(@container, @registry_locker, @reader) if registry_locker.nil?
+      lease = acquire_lease(@registry_locker)
+      # ~ Workaround
+
+      if(registry.nil?)
+        registry_hash = create_registry(candidate_blobs)
+      else
+        registry_hash = load_registry
+      end #if
+        
+      # Pick up the next candidate
+      picked_blob = nil
+      candidate_blobs.each { |candidate_blob|
+        picked_blob = nil
+        start_index = 0
+
+        registry_item = registry_hash[candidate_blob.name]
```

Comment:
:-) Sure. I saved that piece for the next iteration but I guess I'll deal with it now since **someone** asked for it already. :-)